File: simple-stack-example-mvvm-fragments/src/main/java/com/zhuinden/simplestackexamplemvvm/util/ServiceProvider.java
package com.zhuinden.simplestackexamplemvvm.util;

import android.support.annotation.NonNull;

import com.zhuinden.simplestack.ScopeKey;
import com.zhuinden.simplestack.ScopedServices;

public class ServiceProvider
        implements ScopedServices {
    @Override
    public void bindServices(@NonNull ServiceBinder serviceBinder) {
        Object key = serviceBinder.getKey();
        if(key instanceof HasServices) {
            ((HasServices) key).bindServices(serviceBinder);
        }
    }

    public interface HasServices
            extends ScopeKey {
        void bindServices(@NonNull ServiceBinder serviceBinder);
    }
}

File: samples/legacy-architecture-samples/simple-stack-example-mvp-view/src/main/java/com/zhuinden/simplestackdemoexamplemvp/util/scoping/ServiceProvider.kt
package com.zhuinden.simplestackdemoexamplemvp.util.scoping

import com.zhuinden.simplestack.ScopeKey
import com.zhuinden.simplestack.ScopedServices
import com.zhuinden.simplestack.ServiceBinder
import com.zhuinden.simplestackdemoexamplemvp.core.navigation.ViewKey

class ServiceProvider: ScopedServices {
    interface HasServices: ScopeKey {
        fun bindServices(serviceBinder: ServiceBinder)
    }

    override fun bindServices(serviceBinder: ServiceBinder) {
        val key = serviceBinder.getKey<ViewKey>()
        if(key is HasServices) {
            key.bindServices(serviceBinder)
        }
    }
}